packages/next-on-pages/templates/_worker.js/utils/fetch.ts

```diff
@@ -1,79 +1,117 @@
-									const b = Buffer.from(binaryContent);
-									controller.enqueue(b);
-									controller.close();
-									return pump();
-								}
-							},
-						});
-					},
-					async text() {
-						const b = Buffer.from(binaryContent);
-						return b.toString();
-					},
-					async json() {
-						const b = Buffer.from(binaryContent);
-						return JSON.stringify(b.toString());
-					},
-					async blob() {
-						return new Blob(binaryContent);
-					},
-				} as Response;
+/**
+ * This function checks if a given request is trying to fetch an inline if it is it returns a response containing a stream for the asset,
+ * otherwise returns null (signaling that the request hasn't been handled).
+ *
+ * This is necessary so that users can fetch urls such as: `new URL("file", import.meta.url)`
+ * (used for example with `@vercel/og`)
+ *
+ * Note: this function's aim is to mimic the following Next behavior:
+ * 	https://github.com/vercel/next.js/blob/6705c803021d3bdea7fec20e5d98f6899e49836d/packages/next/src/server/web/sandbox/fetch-inline-assets.ts
+ *
+ * @param request the request to handle
+ * @returns the response to return to the caller if the request was for an inline asset one (and the file exists), null otherwise
+ */
+async function handleInlineAssetRequest(request: Request) {
+	if (request.url.startsWith('blob:')) {
+		try {
+			const url = new URL(request.url);
+			const binaryContent = (
+				await import(`./__next-on-pages-dist__/assets/${url.pathname}.bin`)
+			).default;
 
-				// Note: clone is necessary so that body does work
-				resp.clone = (): Response => {
-					return resp;
-				};
+			// Note: we can't generate a real Response object here because this fetch might be called
+			//       at the top level of a dynamically imported module, and such cases produce the following
+			//       error:
+			//           Some functionality, such as asynchronous I/O, timeouts, and generating random values,
+			//           can only be performed while handling a request
+			//       this is a somewhat known workerd behavior (currently kept for security and performance reasons)
+			//
+			//       if the above issue/constraint were to change we should replace the following with a real Response object
+			const resp = {
+				async arrayBuffer() {
+					return binaryContent;
+				},
+				get body(): ReadableStream<unknown> | null {
+					return new ReadableStream({
+						start(controller) {
+							const b = Buffer.from(binaryContent);
+							controller.enqueue(b);
+							controller.close();
+						},
+					});
+				},
+				async text() {
+					const b = Buffer.from(binaryContent);
+					return b.toString();
+				},
+				async json() {
+					const b = Buffer.from(binaryContent);
+					return JSON.stringify(b.toString());
+				},
+				async blob() {
+					return new Blob(binaryContent);
+				},
+			} as Response;
 
+			// Note: clone is necessary so that body does work
+			resp.clone = (): Response => {
 				return resp;
```

Comment:
nit: should we at least spread this into a new object so that it does indeed clone?

```suggestion
				return { ...resp };
```

Reply:
sure 🙂 

it's not really going to be really cloning as the developer could still mutate nested fields, etc...

but anyways I think that this should be enough as I don't even expect this to get actually used (I've just added it for the same of completeness as Greg suggested)